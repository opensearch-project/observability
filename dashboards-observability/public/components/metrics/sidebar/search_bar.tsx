/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiGlobalToastList, EuiSearchBar, EuiToast } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { pplServiceRequestor } from '../helpers/utils';

interface ISearchBarProps {
  allAvailableMetrics: any;
  handleAddMetric: any;
}

export const SearchBar = (props: ISearchBarProps) => {
  const { allAvailableMetrics, handleAddMetric } = props;

  const [query, setQuery] = useState('');
  const [error, setError] = useState(null);

  const [toasts, setToasts] = useState([]);
  const addToast = (res: string) => {
    if (res === 'success') {
      const toast = {
        id: 'success',
        title: 'Metric successfully added!!',
        color: 'success',
      };
      setToasts(toasts.concat(toast));
    } else {
      const toast = {
        id: 'fail',
        title: 'Metric not found.',
        color: 'danger',
      };
      setToasts(toasts.concat(toast));
    }
  };
  const removeToast = (removedToast: any) => {
    setToasts(toasts.filter((toast: any) => toast.id !== removedToast.id));
  };

  const onChange = ({ query }) => {
    const metric = allAvailableMetrics.find((row: any) => row.name.includes(query.text));
    if (metric) {
      handleAddMetric(metric);
      addToast('success');
    } else {
      addToast('fail');
    }
  };

  return (
    <div>
      <EuiSearchBar defaultQuery={''} onChange={onChange} />
      <EuiGlobalToastList toasts={toasts} dismissToast={removeToast} toastLifeTimeMs={6000} />
    </div>
  );
};
