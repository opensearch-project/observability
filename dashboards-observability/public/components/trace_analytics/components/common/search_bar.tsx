/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiSuperDatePicker,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useState } from 'react';
import { uiSettingsService } from '../../../../../common/utils';
import { Filters, FiltersProps } from './filters/filters';

export const renderDatePicker = (
  startTime: string,
  setStartTime: (startTime: string) => void,
  endTime: string,
  setEndTime: (endTime: string) => void
) => {
  return (
    <EuiSuperDatePicker
      start={startTime}
      end={endTime}
      dateFormat={uiSettingsService.get('dateFormat')}
      showUpdateButton={false}
      onTimeChange={(e) => {
        setStartTime(e.start);
        setEndTime(e.end);
      }}
    />
  );
};

export interface SearchBarProps extends FiltersProps {
  query: string;
  setQuery: (query: string) => void;
  startTime: string;
  setStartTime: (startTime: string) => void;
  endTime: string;
  setEndTime: (endTime: string) => void;
}

interface SearchBarOwnProps extends SearchBarProps {
  refresh: () => void;
  page: 'dashboard' | 'traces' | 'services' | 'app';
  datepickerOnly?: boolean;
}

export function SearchBar(props: SearchBarOwnProps) {
  // use another query state to avoid typing delay
  const [query, setQuery] = useState(props.query);
  const setGlobalQuery = _.debounce((q) => props.setQuery(q), 50);

  return (
    <>
      <EuiFlexGroup gutterSize="s">
        {!props.datepickerOnly && (
          <EuiFlexItem>
            <EuiFieldSearch
              fullWidth
              isClearable={false}
              placeholder="Trace ID, trace group name, service name"
              data-test-subj="search-bar-input-box"
              value={query}
              onChange={(e) => {
                setQuery(e.target.value);
                setGlobalQuery(e.target.value);
              }}
              onSearch={props.refresh}
            />
          </EuiFlexItem>
        )}
        <EuiFlexItem grow={false} style={{ maxWidth: '40vw' }}>
          {renderDatePicker(props.startTime, props.setStartTime, props.endTime, props.setEndTime)}
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            data-test-subj="superDatePickerApplyTimeButton"
            data-click-metric-element="trace_analytics.refresh_button"
            iconType="refresh"
            onClick={props.refresh}
          >
            Refresh
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>

      {!props.datepickerOnly && (
        <>
          <EuiSpacer size="s" />
          <Filters
            page={props.page}
            filters={props.filters}
            setFilters={props.setFilters}
            appConfigs={props.appConfigs}
            mode={props.mode}
          />
        </>
      )}
    </>
  );
}
