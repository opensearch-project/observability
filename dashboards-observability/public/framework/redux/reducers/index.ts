/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { combineReducers } from 'redux';

import queriesReducer from '../../../components/event_analytics/redux/slices/query_slice';
import queryResultsReducer from '../../../components/event_analytics/redux/slices/query_result_slice';
import queryTabReducer from '../../../components/event_analytics/redux/slices/query_tab_slice';
import FieldsReducer from '../../../components/event_analytics/redux/slices/field_slice';
import countDistributionReducer from '../../../components/event_analytics/redux/slices/count_distribution_slice';
import explorerVisualizationReducer from '../../../components/event_analytics/redux/slices/visualization_slice';
import explorerVisualizationConfigReducer from '../../../components/event_analytics/redux/slices/viualization_config_slice';
import patternsReducer from '../../../components/event_analytics/redux/slices/patterns_slice';
import metricsReducers from '../../../components/metrics/redux/slices/metrics_slice';

const rootReducer = combineReducers({
  // explorer reducers
  queries: queriesReducer,
  queryResults: queryResultsReducer,
  explorerTabs: queryTabReducer,
  fields: FieldsReducer,
  countDistribution: countDistributionReducer,
  explorerVisualization: explorerVisualizationReducer,
  explorerVisualizationConfig: explorerVisualizationConfigReducer,
  patterns: patternsReducer,
  metrics: metricsReducers,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
