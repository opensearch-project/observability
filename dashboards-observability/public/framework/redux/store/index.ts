/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configureStore } from '@reduxjs/toolkit';
import rootReducer from '../reducers';
import storage from 'redux-persist/lib/storage/session';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';

const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
  enhancers: [],
});

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./rootReducer', () => {
    store.replaceReducer(persistReducer(persistConfig, require('./rootReducer').default));
  });
}

export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);

export default store;
