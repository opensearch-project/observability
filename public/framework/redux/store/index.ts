/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import rootReducer from '../reducers';
import { configureStore } from '@reduxjs/toolkit';

const store = configureStore(
  {
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
    devTools: process.env.NODE_ENV !== 'production',
    enhancers: [],
  }
);

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./rootReducer', () => {
    const newRootReducer = require('./rootReducer').default
    store.replaceReducer(newRootReducer)
  })
}

export type AppDispatch = typeof store.dispatch;

export default store;
