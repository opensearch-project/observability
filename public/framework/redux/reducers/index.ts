/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { combineReducers } from 'redux';

import queriesReducer from '../../../components/explorer/slices/querySlice';
import queryResultsReducer from '../../../components/explorer/slices/queryResultSlice';
import queryTabReducer from '../../../components/explorer/slices/queryTabSlice';
import FieldsReducer from '../../../components/explorer/slices/fieldSlice';
import countDistributionReducer from '../../../components/explorer/slices/countDistributionSlice';
import explorerVisualizationReducer from '../../../components/explorer/slices/visualizationSlice';

const rootReducer = combineReducers({

  // explorer reducers
  queries: queriesReducer,
  queryResults: queryResultsReducer,
  explorerTabs: queryTabReducer,
  fields: FieldsReducer,
  countDistribution: countDistributionReducer,
  explorerVisualization: explorerVisualizationReducer

});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;