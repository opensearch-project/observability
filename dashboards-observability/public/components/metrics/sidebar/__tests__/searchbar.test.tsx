/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { SearchBar } from '../search_bar';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../../../framework/redux/reducers';

describe('Search Bar Component', () => {
  configure({ adapter: new Adapter() });
  const store = createStore(rootReducer, applyMiddleware(thunk));

  it('Search Side Bar Component with no available metrics', async () => {
    const setSearch = jest.fn();

    const wrapper = mount(
      <Provider store={store}>
        <SearchBar setSearch={setSearch} />
      </Provider>
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('Search Side Bar Component with available metrics', async () => {
    const setSearch = jest.fn();

    const wrapper = mount(
      <Provider store={store}>
        <SearchBar setSearch={setSearch} />
      </Provider>
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
