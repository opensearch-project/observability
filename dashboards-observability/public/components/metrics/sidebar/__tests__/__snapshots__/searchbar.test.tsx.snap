// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar Component Search Side Bar Component with available metrics 1`] = `
<SearchBar
  allAvailableMetrics={
    Array [
      Object {
        "catalog": "CUSTOM_METRICS",
        "id": "HIlAQYQBiaYaSxpXJ73K",
        "name": "[Prometheus Metric] prometheus.process_resident_memory_bytes",
        "recentlyCreated": true,
        "type": "line",
      },
      Object {
        "catalog": "CUSTOM_METRICS",
        "id": "Y4muP4QBiaYaSxpXk7r8",
        "name": "new metric",
        "recentlyCreated": false,
        "type": "line",
      },
      Object {
        "catalog": "CUSTOM_METRICS",
        "id": "HolAQYQBiaYaSxpXKL0T",
        "name": "[Prometheus Metric] prometheus.go_memstats_heap_sys_bytes",
        "recentlyCreated": true,
        "type": "line",
      },
      Object {
        "catalog": "CUSTOM_METRICS",
        "id": "tomAP4QBiaYaSxpXALls",
        "name": "[Logs] Average ram usage per day by windows os ",
        "recentlyCreated": false,
        "type": "line",
      },
      Object {
        "catalog": "prometheus",
        "id": "prometheus.process_resident_memory_bytes",
        "name": "prometheus.process_resident_memory_bytes",
        "recentlyCreated": false,
        "type": "gauge",
      },
    ]
  }
  handleAddMetric={[MockFunction]}
>
  <div>
    <EuiSearchBar
      defaultQuery=""
      onChange={[Function]}
    >
      <EuiFlexGroup
        alignItems="center"
        gutterSize="m"
        wrap={true}
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
        >
          <EuiFlexItem
            className="euiSearchBar__searchHolder"
            grow={true}
          >
            <div
              className="euiFlexItem euiSearchBar__searchHolder"
            >
              <EuiSearchBox
                incremental={false}
                isInvalid={false}
                onSearch={[Function]}
                placeholder="Search..."
                query=""
              >
                <EuiFieldSearch
                  aria-label="This is a search bar. After typing your query, hit enter to filter the results lower in the page."
                  compressed={false}
                  defaultValue=""
                  fullWidth={true}
                  incremental={false}
                  inputRef={[Function]}
                  isClearable={true}
                  isInvalid={false}
                  isLoading={false}
                  onSearch={[Function]}
                  placeholder="Search..."
                >
                  <EuiFormControlLayout
                    compressed={false}
                    fullWidth={true}
                    icon="search"
                    isLoading={false}
                  >
                    <div
                      className="euiFormControlLayout euiFormControlLayout--fullWidth"
                    >
                      <div
                        className="euiFormControlLayout__childrenWrapper"
                      >
                        <EuiValidatableControl
                          isInvalid={false}
                        >
                          <input
                            aria-label="This is a search bar. After typing your query, hit enter to filter the results lower in the page."
                            className="euiFieldSearch euiFieldSearch--fullWidth"
                            defaultValue=""
                            onKeyUp={[Function]}
                            placeholder="Search..."
                            type="search"
                          />
                        </EuiValidatableControl>
                        <EuiFormControlLayoutIcons
                          compressed={false}
                          icon="search"
                          isLoading={false}
                        >
                          <div
                            className="euiFormControlLayoutIcons"
                          >
                            <EuiFormControlLayoutCustomIcon
                              size="m"
                              type="search"
                            >
                              <span
                                className="euiFormControlLayoutCustomIcon"
                              >
                                <EuiIcon
                                  aria-hidden="true"
                                  className="euiFormControlLayoutCustomIcon__icon"
                                  size="m"
                                  type="search"
                                >
                                  <EuiIconSearch
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      className="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <path
                                        d="M11.271 11.978l3.872 3.873a.502.502 0 00.708 0 .502.502 0 000-.708l-3.565-3.564c2.38-2.747 2.267-6.923-.342-9.532-2.73-2.73-7.17-2.73-9.898 0-2.728 2.729-2.728 7.17 0 9.9a6.955 6.955 0 004.949 2.05.5.5 0 000-1 5.96 5.96 0 01-4.242-1.757 6.01 6.01 0 010-8.486c2.337-2.34 6.143-2.34 8.484 0a6.01 6.01 0 010 8.486.5.5 0 00.034.738z"
                                      />
                                    </svg>
                                  </EuiIconSearch>
                                </EuiIcon>
                              </span>
                            </EuiFormControlLayoutCustomIcon>
                          </div>
                        </EuiFormControlLayoutIcons>
                      </div>
                    </div>
                  </EuiFormControlLayout>
                </EuiFieldSearch>
              </EuiSearchBox>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
    </EuiSearchBar>
    <EuiGlobalToastList
      dismissToast={[Function]}
      side="right"
      toastLifeTimeMs={6000}
      toasts={Array []}
    >
      <div
        aria-live="polite"
        className="euiGlobalToastList euiGlobalToastList--right"
        role="region"
      />
    </EuiGlobalToastList>
  </div>
</SearchBar>
`;

exports[`Search Bar Component Search Side Bar Component with no available metrics 1`] = `
<SearchBar
  allAvailableMetrics={Array []}
  handleAddMetric={[MockFunction]}
>
  <div>
    <EuiSearchBar
      defaultQuery=""
      onChange={[Function]}
    >
      <EuiFlexGroup
        alignItems="center"
        gutterSize="m"
        wrap={true}
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
        >
          <EuiFlexItem
            className="euiSearchBar__searchHolder"
            grow={true}
          >
            <div
              className="euiFlexItem euiSearchBar__searchHolder"
            >
              <EuiSearchBox
                incremental={false}
                isInvalid={false}
                onSearch={[Function]}
                placeholder="Search..."
                query=""
              >
                <EuiFieldSearch
                  aria-label="This is a search bar. After typing your query, hit enter to filter the results lower in the page."
                  compressed={false}
                  defaultValue=""
                  fullWidth={true}
                  incremental={false}
                  inputRef={[Function]}
                  isClearable={true}
                  isInvalid={false}
                  isLoading={false}
                  onSearch={[Function]}
                  placeholder="Search..."
                >
                  <EuiFormControlLayout
                    compressed={false}
                    fullWidth={true}
                    icon="search"
                    isLoading={false}
                  >
                    <div
                      className="euiFormControlLayout euiFormControlLayout--fullWidth"
                    >
                      <div
                        className="euiFormControlLayout__childrenWrapper"
                      >
                        <EuiValidatableControl
                          isInvalid={false}
                        >
                          <input
                            aria-label="This is a search bar. After typing your query, hit enter to filter the results lower in the page."
                            className="euiFieldSearch euiFieldSearch--fullWidth"
                            defaultValue=""
                            onKeyUp={[Function]}
                            placeholder="Search..."
                            type="search"
                          />
                        </EuiValidatableControl>
                        <EuiFormControlLayoutIcons
                          compressed={false}
                          icon="search"
                          isLoading={false}
                        >
                          <div
                            className="euiFormControlLayoutIcons"
                          >
                            <EuiFormControlLayoutCustomIcon
                              size="m"
                              type="search"
                            >
                              <span
                                className="euiFormControlLayoutCustomIcon"
                              >
                                <EuiIcon
                                  aria-hidden="true"
                                  className="euiFormControlLayoutCustomIcon__icon"
                                  size="m"
                                  type="search"
                                >
                                  <EuiIconEmpty
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    />
                                  </EuiIconEmpty>
                                </EuiIcon>
                              </span>
                            </EuiFormControlLayoutCustomIcon>
                          </div>
                        </EuiFormControlLayoutIcons>
                      </div>
                    </div>
                  </EuiFormControlLayout>
                </EuiFieldSearch>
              </EuiSearchBox>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
    </EuiSearchBar>
    <EuiGlobalToastList
      dismissToast={[Function]}
      side="right"
      toastLifeTimeMs={6000}
      toasts={Array []}
    >
      <div
        aria-live="polite"
        className="euiGlobalToastList euiGlobalToastList--right"
        role="region"
      />
    </EuiGlobalToastList>
  </div>
</SearchBar>
`;
